python/lsst/daf/butler/delegates/arrownumpy.py

```diff
@@ -0,0 +1,97 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Support for reading numpy tables (structured arrays) with the Arrow
+formatter.
+"""
+from __future__ import annotations
+
+from typing import Any, Mapping, Optional
+
+import numpy as np
+from lsst.daf.butler import StorageClassDelegate
+from lsst.daf.butler.formatters.parquet import ArrowNumpySchema
+from lsst.utils.introspection import get_full_type_name
+from lsst.utils.iteration import ensure_iterable
+
+__all__ = ["ArrowNumpyDelegate"]
+
+
+class ArrowNumpyDelegate(StorageClassDelegate):
+    def getComponent(self, composite: np.ndarray, componentName: str) -> Any:
+        """Get a component from a numpy table stored via ArrowNumpy.
+
+        Parameters
+        ----------
+        composite : `~numpy.ndarray`
+            Numpy table to access component.
+        componentName : `str`
+            Name of component to retrieve.
+
+        Returns
+        -------
+        component : `object`
+            The component.
+
+        Raises
+        ------
+        AttributeError
+            The component can not be found.
+        """
+        if componentName == "columns":
+            return list(composite.dtype.names)
+        elif componentName == "schema":
+            return ArrowNumpySchema(composite.dtype)
+        elif componentName == "rowcount":
+            return len(composite)
+        else:
+            raise AttributeError(
+                f"Do not know how to retrieve component {componentName} from {get_full_type_name(composite)}"
+            )
+
+    def handleParameters(
```

Comment:
There's an uncomfortable amount of code duplication between this method and the ones in the other delegates. Seems like only one line differs (the "if column not in") and the instance check (the python type for that could be a class attribute). I don't like that much repeated logic across multiple files.